### ISSUE TITLE
Perfview dump size doesn't match the real dump size.

### ISSUE THREAD

**Issue by OnurGumus** [microsoft/perfview#160]
I have a managed memory dump with 15 GB. When I open this dump with Perfview (despite I select 1 Billion objects) , and include unreachable memory, it totals around 12 GB. Within WinDBG if I run !eeheap -gc  I can see managed heap size is 15 gigs in deed. Why this difference is happening ? Is it possible to get a perfect snapshot of the heap from an existing dump by using perfview ?

**Comment by vancem** [comment]
After converting the .dmp file open the log (log button in lower right corner), and it will probably say that the log was too big and refer you to a file in %TEMP% for the complete log file.   This is the interesting file.   Attach that to this issue (drag and drop works).    This details exactly what pefVIew found when it looked at the heap.  

You should also include the data from the eeheap -gc that shows a different number.

**Comment by OnurGumus** [comment]
Below is the heap graph via clr profiler after running !traverseheap via windbg, showing 13 Gigs ( in addition to that 2 gigs is garbage and free so it adds up to 15 Gigs)
![perfview_heapgraph](https://cloud.githubusercontent.com/assets/4114592/23202358/442ea75a-f8e6-11e6-8796-3b0981f0ff91.png)

Below is the initial config  before running:

![perfview1](https://cloud.githubusercontent.com/assets/4114592/23203067/93ba4646-f8e9-11e6-9d35-72e2882be1dd.png)

Here's screenshot from perfview:

![perfviewresult](https://cloud.githubusercontent.com/assets/4114592/23203049/7d7e5a02-f8e9-11e6-9c80-e33fa211a46f.png)


And here's the log file:

[perfview.txt](https://github.com/Microsoft/perfview/files/792651/perfview.txt)

And finally output of !eeheap -gc
[eeheap.txt](https://github.com/Microsoft/perfview/files/792574/eeheap.txt)

**Comment by vancem** [comment]
Thanks @OnurGumus

It looks like the PerfView.txt file is a cut and past of the log window.  You will see at the very top of this text is a line that indicates that it is a truncated view and where to find the full log file.   Could you attach that instead.

**Comment by OnurGumus** [comment]
[PerfViewLogFile.zip](https://github.com/Microsoft/perfview/files/799344/PerfViewLogFile.zip)

Hello @vancem , sorry for the delay. Here's the full log zipped as you requested.

**Comment by OnurGumus** [comment]
@vancem any updates?

**Comment by vancem** [comment]
There is no easy fix your issue.   Basically the representation of the GC heap uses 32 bit numbers for 'pointers' into its heap.   This places a limit on how many objects can be in this representation.    There are certainly solutions to this, but it requires some thought as to the tradeoffs (hitting this limit is pretty rare, so making the common case pay for this uncommon case would be unfortunate).  

I have some thoughts on this (allow pointers to point directly to the object up to 2GB into the heap, and then after that index to a table (so you have 2B more pointers that can be 64 bits).    It will not be a trivial change (but hopefully not too big).   Realistically we need to figure out how big of a problem this is, and just how big of ha change it is, before really investing in it.  

Note that generally speaking, the heap you get is a sampling of the GC heap, and doing analysis on it is likely to be pretty accurate (especially if the true size of the heap is < 2X the maximum).   So yes, it is not a perfect trace, but it probably still gives you most of the value.

**Comment by OnurGumus** [comment]
However from the logs as you see there are only 100+ million objects stored. Not 2B.  I manually try to increment the m_maxNodeCount and it works! Then the FastSerializer throws OutOfMemoryException explicitly
https://github.com/Microsoft/perfview/blob/36ac95461697412b02078cb91aefde53701d208a/src/FastSerialization/StreamReaderWriter.cs
in MakeSpace() method. Due to max array size exceeded.

The dumps I analyze is usually greater than 15 GB. And my manager usually ask me what's the memory usage of a particular type. Unfortunately I cannot answer the question by looking to perfview. (But only either by clrprofiler or !dumpheap -stat -live )

**Comment by vancem** [comment]
Yes, it is not the number of objects that is limited, it is the size of the GC heap representation.    This is why this is non-trivial to fix.   

It is true that without some fix, you won't be able to say how many bytes a particular type uses.   However you will know it for chunk that was decoded, and thus you know the RELATIVE cost per type, which is what you probably care about more.   You can also estimate the total amount by scaling appropriately (but typically you don't care, the percentages are enough).   

Thus if you just ignored the fact that it is not the whole heap, you probably be doing the 'right thing' (you would know what things costs what in the program and you can optimize it).

**Comment by OnurGumus** [comment]
@vancem  Thank you for your insightful comments. I now have a better picture about the problem and how should I use perfview. Since there is nothing else to do about it at the moment I am closing this issue.

**Comment by OnurGumus** [comment]
Hello, I am reopening the issue because I have another case which is not consistent about what you say. The total dumpsize is 12 GB. Perfview shows total metric as 8.5 GB with %99.3 as unreachable. CLR Profiler (after !traverseheap) Shows 3.9 GB reachable data. I believe this partial snapshot perfview is taking may not be good at sampling. So perhaps it walks all through the unused part of the memory. Again perhaps you should provide an option for  perfview to ignore unreachable memory when walking on the crash dumps.

**Comment by vancem** [comment]
When heap viewing was first added to PerfView, I had hoped that 'random' sampling of the heap would work well enough.    We experimented with a variety of possibilities, but ultimately the issue is that while sampling makes sense, when you take a sample you also need to keep all the object that make it reachable.    However to that efficiently on many samples, you basically need the 'pointed to' graph, which is a structure that is basically as large as representation of the heap itself.   

Certainly the best answer is to fix it so that the representation does not have limitations that you are hitting.    We can keep this bug open as a marker to do that, but it may not happen for a while.  

Note that partial heaps are pretty much guaranteed to have many 'falsely' unreachable object (after all any objects pointed to only from that part that was not traversed will seem unreachable).     However I have also found that reachability is really not as important as you might think.   What is true is that the GC does convert objects to a special 'Free' type, when it actually collects them, which means that although things may be unreachable in the view, if they have a type, they are probably reachable.   

It is true that you might not know the complete path that makes them reachable, but for big things there are bound to be cases that do have paths, and that is often enough to figure out the problem.

**Comment by sharwell** [comment]
@vancem Right shifting everything that points *to a reference type* by 3 bits will allow you to address 32GiB without overflowing a 32-bit integer. Then use 64-bit pointers for managed pointers to value types.

**Comment by vancem** [comment]
The structure that is overflowing is a big blob that PerfView's 'Graphs.Graph' type uses to represent a single node in the graph.   This structure has a 4 byte 'pointer' that points into that blob.   The things in the blog are designed to take up as little space as possible so the representations in this blob might take any number of bytes (often just a few).    Thus we can't play tricks that take advantage of alignment of these values because they are not aligned.  

There is of course the obvious solution of using a larger pointer (e.g. 8 byte pointer), into the blob to represent a node.   This change will increase the size of the represenation by probably 25%. 

I am not averse to fiddling with the code (it should not take much) to allow this to be decided by a #ifdef at compile time.    This is OK by me because it changes the code very little and is conceptually very simple.   What I don't like about it is that it is not very pay-for play (the current situation is arguably not that bad, but its solution requires inefficiency for the 99% case).     This is why I am biased toward a solution that only 'expands' when it is needed.   The question is can this be done without increasing the complexity of the code in an interesting way (I *think* it can).  

My main point is that it requires at least a little thought/experimentation before we make the decision, and I would like to be involved early before people invest a whole lot.  Doing what I suggest above (expanding the index into the blob to a long) is simple enough that someone could go and do it, and we could do experiments with it until we determine if that is 'good enough' or we want to make something more pay-for-play.

**Comment by brianrob** [comment]
Closing due to inactivity.
